php-checks/src/main/java/org/sonar/php/checks/UnusedLocalVariableCheck.java

```diff
@@ -19,144 +19,115 @@
-      PHPGrammar.LIST_EXPR,
-      PHPGrammar.FOREACH_EXPR);
-  }
+  public static final String KEY = "S1481";
+  private static final String MESSAGE = "Remove this unused \"%s\" local variable.";
 
-  @Override
-  public void leaveFile(@Nullable AstNode astNode) {
-    scopes.clear();
-  }
+  private List<IdentifierTree> exclusions = new ArrayList<>();
 
   @Override
-  public void visitNode(AstNode astNode) {
-    if (astNode.is(FunctionUtils.functions())) {
-      scopes.push(new LocalVariableScope());
-      getCurrentScope().declareParameters(astNode);
-    } else if (!scopes.isEmpty()) {
-
-      if (astNode.is(PHPGrammar.GLOBAL_STATEMENT)) {
-        getCurrentScope().declareGlobals(astNode);
-
-      } else if (astNode.is(PHPGrammar.STATIC_STATEMENT)) {
-        getCurrentScope().declareStaticVariables(astNode);
-
-      } else if (astNode.is(PHPGrammar.LEXICAL_VAR_LIST)) {
-        getCurrentScope().declareLexicalVariable(astNode, getOuterScope());
-
-      } else if (astNode.is(PHPGrammar.FOREACH_EXPR)) {
-        getCurrentScope().declareForeachVariable(astNode);
-
-      } else if (astNode.is(PHPGrammar.VAR_IDENTIFIER) && !isDeclaration(astNode)) {
-        getCurrentScope().useVariable(astNode);
+  public void visitFunctionExpression(FunctionExpressionTree tree) {
+    if (tree.lexicalVars() != null) {
+      Scope parentScope = context().symbolTable().getScopeFor(tree).outer();
 
-      } else if (astNode.is(PHPGrammar.SEMI_COMPLEX_ENCAPS_VARIABLE)) {
-        getCurrentScope().useVariale("$" + astNode.getFirstChild(PHPGrammar.EXPRESSION).getTokenOriginalValue());
+      for (VariableTree variableTree : tree.lexicalVars().variables()) {
 
-      } else if (astNode.is(PHPGrammar.ASSIGNMENT_EXPR, PHPGrammar.ASSIGNMENT_BY_REFERENCE)) {
-        declareNewLocalVariable(astNode);
+        if (variableTree.is(Tree.Kind.VARIABLE_IDENTIFIER)) {
+          VariableIdentifierTree variableIdentifier = (VariableIdentifierTree) variableTree;
+          Symbol parentScopeSymbol = parentScope.getSymbol(variableIdentifier.text());
 
-      } else if (astNode.is(PHPGrammar.LIST_EXPR)) {
-        getCurrentScope().declareListVariable(astNode);
+          if (parentScopeSymbol == null) {
+            exclusions.add(variableIdentifier);
+          }
+        }
       }
     }
-  }
 
-  private boolean isDeclaration(AstNode varIdentifier) {
-    AstNode parent = varIdentifier.getParent();
-    return parent.getParent().is(PHPGrammar.GLOBAL_VAR) || parent.is(PHPGrammar.STATIC_VAR, PHPGrammar.LEXICAL_VAR);
+    super.visitFunctionExpression(tree);
   }
 
   @Override
-  public void leaveNode(AstNode astNode) {
-    if (astNode.is(FunctionUtils.functions())) {
-      reportUnusedVariable();
-      scopes.pop();
+  public void visitFunctionCall(FunctionCallTree tree) {
+    for (ExpressionTree argument : tree.arguments()) {
+      if (argument.is(Tree.Kind.VARIABLE_IDENTIFIER)) {
+        exclusions.add((IdentifierTree) argument);
```

Comment:
@pynicolas by this way we ignore symbols declared as function call argument. See this case https://github.com/SonarCommunity/sonar-php/blob/cac74ad10283398e9fe3708c7dfd955f5885045f/php-checks/src/test/resources/checks/UnusedLocalVariableCheck.php#L200 (`arg1` symbol is created here and not used, but you are obliged to pass it as argument).